[UnifiedPDF] Data detection overlays should show on PDFs

aprotyas opened the pull request:
#### 37723d9e6498a980cb60caec159f92c546f4726b
<pre>
[UnifiedPDF] Data detection overlays should show on PDFs
<a href="https://bugs.webkit.org/show_bug.cgi?id=270998">https://bugs.webkit.org/show_bug.cgi?id=270998</a>
<a href="https://rdar.apple.com/123761050">rdar://123761050</a>

Reviewed by Simon Fraser.

Currently, we do not surface any data detection results on
UnifiedPDFPlugin. This patch introduces the plumbing required to
fetch data detection results from PDFKit, the infrastructure to display
data detector overlays, and the requisite event handling/state
management to only display (or act on) a data detection result when the
mouse hovers/clicks over it.

To facilitate this end:

- Introduce two classes WebKit::PDFDataDetectorItem and
  WebKit::PDFDataDetectorOverlayController. The former encapsulates all
  the data we need to represent a data detection scanning result
  obtained from PDFKit, while the latter manages all the state for us to
  be able to indicate data detector results as page overlays.

- Introduce new public helpers on UnifiedPDFPlugin that
  PDFDataDetectorOverlayController can use to delegate work. Notably,
  this is required for us to implement the updated
  DataDetectorHighlightClient interface from 276060@main.

- Modify WebCore::DataDetectorHighlight to understand a new type of
  highlight corresponding to PDF selections.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Add a new PlatformEnable flag so we don&apos;t have to litter the codebase
with #if ENABLE(UNIFIED_PDF) &amp;&amp; ENABLE(DATA_DETECTION) &amp;&amp; PLATFORM(MAC)

* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.mm:

DDResultIsPastDate is available (and used) on non-IOS_FAMILY platforms,
so let&apos;s move its declaration out from the platform-specific section.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mac/DataDetectorHighlight.h:
* Source/WebCore/platform/mac/DataDetectorHighlight.mm:
(WebCore::DataDetectorHighlight::createForPDFSelection):
(WebCore::DataDetectorHighlight::isRangeSupportingType const):
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
(WebKit::PDFPluginBase::mousePositionInView const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.h: Added.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.mm: Added.
(WebKit::hasActionsForResult):
(WebKit::resultIsPastDate):
(WebKit::PDFDataDetectorItem::create):
(WebKit::PDFDataDetectorItem::PDFDataDetectorItem):
(WebKit::PDFDataDetectorItem::scannerResult const):
(WebKit::PDFDataDetectorItem::hasActions const):
(WebKit::PDFDataDetectorItem::isPastDate const):
(WebKit::PDFDataDetectorItem::selection const):
(WebKit::PDFDataDetectorItem::pageBounds const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h: Added.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm: Added.
(WebKit::PDFDataDetectorOverlayController::PDFDataDetectorOverlayController):
(WebKit::PDFDataDetectorOverlayController::protectedPlugin const):
(WebKit::PDFDataDetectorOverlayController::installOverlayIfNeeded):
(WebKit::PDFDataDetectorOverlayController::uninstallOverlay):
(WebKit::PDFDataDetectorOverlayController::teardown):
(WebKit::createPlatformDataDetectorHighlight):
(WebKit::PDFDataDetectorOverlayController::createPlatformDataDetectorHighlight const):
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::handleDataDetectorAction):
(WebKit::PDFDataDetectorOverlayController::updateDataDetectorHighlightsIfNeeded):
(WebKit::PDFDataDetectorOverlayController::updatePlatformHighlightData):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
(WebKit::PDFDataDetectorOverlayController::willMoveToPage):
(WebKit::PDFDataDetectorOverlayController::scheduleRenderingUpdate):
(WebKit::PDFDataDetectorOverlayController::deviceScaleFactor const):
(WebKit::PDFDataDetectorOverlayController::createGraphicsLayer):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::enableDataDetection):
(WebKit::UnifiedPDFPlugin::handleClickForDataDetectionResult):
(WebKit::UnifiedPDFPlugin::createGraphicsLayer):

Refactor createGraphicsLayer into three overloads so it isn&apos;t
inconvenient to use internally or from the overlay controller.

(WebKit::UnifiedPDFPlugin::scheduleRenderingUpdate):
(WebKit::UnifiedPDFPlugin::pageAtIndex const):

Expose this helper to avoid having to pass the entire PDFDocumentLayout
object to the overlay controller.

(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

Repaint the data detector overlays to our hearts content on a scroll. To
be optimized!

(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::rectForSelectionInPluginSpace const):
(WebKit::UnifiedPDFPlugin::installDataDetectorOverlay):
(WebKit::UnifiedPDFPlugin::uninstallDataDetectorOverlay):

Canonical link: <a href="https://commits.webkit.org/276131@main">https://commits.webkit.org/276131@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/78f9f4454360872bf3ebfa7b197703d5a3b441e0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/43843 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/22891 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/46274 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/46486 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/39931 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/46147 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/26825 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/20293 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/36183 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/44417 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/19924 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/37764 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/17181 "Passed tests") | 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/17446 "Unexpected infrastructure issue, retrying build") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/38844 "Passed tests") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/1896 "Built successfully") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/37248 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/40018 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/39145 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/48046 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/43451 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/18859 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/15437 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/42994 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/20253 "Built successfully") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/41690 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9749 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/20456 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/50472 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/19877 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/10184 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

aprotyas commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.mm:
```diff
@@ -0,0 +1,95 @@
+#include <wtf/cf/TypeCastsCF.h>
+#include <wtf/text/WTFString.h>
+
+#include "PDFKitSoftlink.h"
```
Inevitable style guide confusion here because of two conflicting rules:

1. `Bad include order. Mixing system and custom headers.  [build/include_order] [4]`
2. `*SoftLink.h header should be included after all other headers.  [build/include_order] [4]`

aprotyas replied:
This seems like a style-checker bug, because separating the `#include "PDFKitSoftlink.h` line into its own paragraph still confuses the checker.

aprotyas replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=271014

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.h:
```diff
@@ -0,0 +1,74 @@
+    bool hasActions() const;
+    bool isPastDate() const;
+    RetainPtr<PDFSelection> selection() const;
+    FloatRect pageBounds() const;
```
Is this just the bounds of the page that the item is in? Can we just store a page index instead, and get the bounds from PDFDocumentLayout?

aprotyas replied:
Yes, just the bounds. I don't need to expose this right now, so I'll remove it. When we do need this, I'll store a page index instead and query the PDFDocumentLayout with it.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
```diff
@@ -0,0 +1,114 @@
+    explicit PDFDataDetectorOverlayController(UnifiedPDFPlugin&);
+    virtual ~PDFDataDetectorOverlayController() = default;
+
+    WebCore::DataDetectorHighlight* wowow1() const { return activeHighlight(); }
```
OOPS

aprotyas replied:
Removed!

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h:
```diff
@@ -0,0 +1,114 @@
+
+    enum class ShouldUpdatePlatformHighlightData : bool { No, Yes };
+    void didInvalidateHighlightOverlayRects(ShouldUpdatePlatformHighlightData = ShouldUpdatePlatformHighlightData::Yes);
+    void teardown();
```
Put this up next to the destructor.

aprotyas replied:
Fixed!

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -0,0 +1,292 @@
+    auto selectionBoundsInPluginSpace = plugin->rectForSelectionInPluginSpace(dataDetectorItem.selection().get());
+    auto visibleContentRectInPluginSpace = plugin->visibleContentRect();
+
+    return ::WebKit::createPlatformDataDetectorHighlight(Vector<FloatRect>::from(WTFMove(selectionBoundsInPluginSpace)), WTFMove(visibleContentRectInPluginSpace));
```
The `::WebKit` is necessary?

aprotyas replied:
Yes, only to disambiguate between the free function and the class method. I could also just rename the free function `createPlatformDataDetectorHighlightImpl` or something.

aprotyas replied:
Keeping it as-is for now.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -0,0 +1,292 @@
+
+    auto rectForSelectionInPluginSpace = roundedIntRect(plugin->rectForSelectionInPluginSpace(dataDetectorItem.selection().get()));
+
+    plugin->handleClickForDataDetectionResult({ dataDetectorItem.scannerResult(), WTFMove(rectForSelectionInPluginSpace) }, WTFMove(mousePositionInPluginSpace));
```
Dunno if it's worth moving rects and points.

aprotyas replied:
Yes, I suppose it doesn't really help. I'll remove these moves.

smfr commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm:
```diff
@@ -0,0 +1,292 @@
+
+                    Ref coreHighlight = DataDetectorHighlight::createForPDFSelection(*this, createPlatformDataDetectorHighlight(dataDetectorItem.get()));
+
+                    return std::optional { std::make_pair(WTFMove(dataDetectorItem), WTFMove(coreHighlight)) };
```
Is the `std::optional` necessary?

aprotyas replied:
Not required. Removed!

aprotyas committed:
[UnifiedPDF] Data detection overlays should show on PDFs
https://bugs.webkit.org/show_bug.cgi?id=270998
rdar://123761050

Reviewed by Simon Fraser.

Currently, we do not surface any data detection results on
UnifiedPDFPlugin. This patch introduces the plumbing required to
fetch data detection results from PDFKit, the infrastructure to display
data detector overlays, and the requisite event handling/state
management to only display (or act on) a data detection result when the
mouse hovers/clicks over it.

To facilitate this end:

- Introduce two classes WebKit::PDFDataDetectorItem and
  WebKit::PDFDataDetectorOverlayController. The former encapsulates all
  the data we need to represent a data detection scanning result
  obtained from PDFKit, while the latter manages all the state for us to
  be able to indicate data detector results as page overlays.

- Introduce new public helpers on UnifiedPDFPlugin that
  PDFDataDetectorOverlayController can use to delegate work. Notably,
  this is required for us to implement the updated
  DataDetectorHighlightClient interface from 276060@main.

- Modify WebCore::DataDetectorHighlight to understand a new type of
  highlight corresponding to PDF selections.

* Source/WTF/wtf/PlatformEnableCocoa.h:

Add a new PlatformEnable flag so we don't have to litter the codebase
with #if ENABLE(UNIFIED_PDF) && ENABLE(DATA_DETECTION) && PLATFORM(MAC)

* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/DataDetectorsCoreSoftLink.mm:

DDResultIsPastDate is available (and used) on non-IOS_FAMILY platforms,
so let's move its declaration out from the platform-specific section.

* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/platform/mac/DataDetectorHighlight.h:
* Source/WebCore/platform/mac/DataDetectorHighlight.mm:
(WebCore::DataDetectorHighlight::createForPDFSelection):
(WebCore::DataDetectorHighlight::isRangeSupportingType const):
* Source/WebKit/Platform/spi/Cocoa/PDFKitSPI.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.h:
* Source/WebKit/Shared/Cocoa/PDFKitSoftLink.mm:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.h:
* Source/WebKit/WebProcess/Plugins/PDF/PDFPluginBase.mm:
(WebKit::PDFPluginBase::lastKnownMousePositionInView const):
(WebKit::PDFPluginBase::mousePositionInView const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.h: Added.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorItem.mm: Added.
(WebKit::hasActionsForResult):
(WebKit::resultIsPastDate):
(WebKit::PDFDataDetectorItem::create):
(WebKit::PDFDataDetectorItem::PDFDataDetectorItem):
(WebKit::PDFDataDetectorItem::scannerResult const):
(WebKit::PDFDataDetectorItem::hasActions const):
(WebKit::PDFDataDetectorItem::isPastDate const):
(WebKit::PDFDataDetectorItem::selection const):
(WebKit::PDFDataDetectorItem::pageBounds const):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.h: Added.
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/PDFDataDetectorOverlayController.mm: Added.
(WebKit::PDFDataDetectorOverlayController::PDFDataDetectorOverlayController):
(WebKit::PDFDataDetectorOverlayController::protectedPlugin const):
(WebKit::PDFDataDetectorOverlayController::installOverlayIfNeeded):
(WebKit::PDFDataDetectorOverlayController::uninstallOverlay):
(WebKit::PDFDataDetectorOverlayController::teardown):
(WebKit::createPlatformDataDetectorHighlight):
(WebKit::PDFDataDetectorOverlayController::createPlatformDataDetectorHighlight const):
(WebKit::PDFDataDetectorOverlayController::handleMouseEvent):
(WebKit::PDFDataDetectorOverlayController::handleDataDetectorAction):
(WebKit::PDFDataDetectorOverlayController::updateDataDetectorHighlightsIfNeeded):
(WebKit::PDFDataDetectorOverlayController::updatePlatformHighlightData):
(WebKit::PDFDataDetectorOverlayController::didInvalidateHighlightOverlayRects):
(WebKit::PDFDataDetectorOverlayController::willMoveToPage):
(WebKit::PDFDataDetectorOverlayController::scheduleRenderingUpdate):
(WebKit::PDFDataDetectorOverlayController::deviceScaleFactor const):
(WebKit::PDFDataDetectorOverlayController::createGraphicsLayer):
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.h:
* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::UnifiedPDFPlugin):
(WebKit::UnifiedPDFPlugin::teardown):
(WebKit::UnifiedPDFPlugin::installPDFDocument):
(WebKit::UnifiedPDFPlugin::enableDataDetection):
(WebKit::UnifiedPDFPlugin::handleClickForDataDetectionResult):
(WebKit::UnifiedPDFPlugin::createGraphicsLayer):

Refactor createGraphicsLayer into three overloads so it isn't
inconvenient to use internally or from the overlay controller.

(WebKit::UnifiedPDFPlugin::scheduleRenderingUpdate):
(WebKit::UnifiedPDFPlugin::pageAtIndex const):

Expose this helper to avoid having to pass the entire PDFDocumentLayout
object to the overlay controller.

(WebKit::UnifiedPDFPlugin::didChangeScrollOffset):

Repaint the data detector overlays to our hearts content on a scroll. To
be optimized!

(WebKit::UnifiedPDFPlugin::handleMouseEvent):
(WebKit::UnifiedPDFPlugin::rectForSelectionInPluginSpace const):
(WebKit::UnifiedPDFPlugin::installDataDetectorOverlay):
(WebKit::UnifiedPDFPlugin::uninstallDataDetectorOverlay):

Canonical link: https://commits.webkit.org/276131@main
